src/common/com/intellij/plugins/haxe/lang/util/HaxeExpressionUtil.java

```diff
@@ -73,4 +86,294 @@ public static boolean isArrayExpression(HaxeExpression expr) {
+   * @return true, if the expression is considered constant; false, otherwise.
+   */
+  public static boolean isConstantExpression(HaxeExpression expr) {
+    return ConstantClass.NOT_CONSTANT != classifyConstantExpression(expr);
+  }
+
+  /**
+   * Resolve the the type of the constant, if it *is* a constant.
+   * @return The ConstantClass, or NULL if the element is not constant.
+   */
+  private static ConstantClass classifyConstantExpression(HaxeExpression expr) {
+    // Protect from circular references.
+    if (constantsInProcess.get().contains(expr)) {
+      return ConstantClass.NOT_CONSTANT;
+    }
+    constantsInProcess.get().add(expr);
+    try {
+
+      ConstantClass cc = innerClassifier(expr);
+      return cc;
+
+    } finally {
+      constantsInProcess.get().remove(expr);
+    }
+  }
+
+  private static ConstantClass innerClassifier(HaxeExpression expr) {
+
+    if (null == expr) return ConstantClass.NOT_CONSTANT;
+
+    // According to Simn:
+    // Functions, arrays, maps, and anonymous objects are NOT considered constant
+    // expressions.  The compiler ensures that "inline field == inline field" always
+    // holds.  If arrays were allowed, e.g. "static inline var myconstant = []",
+    // then "myconstant == myconstant" would come out as "[] == []" and that's false.
+
+    // Compiler's check:
+    //
+    //    let type_function_arg_value ctx t c do_display =
+    //        match c with
+    //            | None -> None
+    //            | Some e ->
+    //                let p = pos e in
+    //                let e = if do_display then Display.ExprPreprocessing.process_expr ctx.com e else e in
+    //                let e = ctx.g.do_optimize ctx (type_expr ctx e (WithType.with_type t)) in
+    //                unify ctx e.etype t p;
+    //                let rec loop e = match e.eexpr with
+    //                    | TConst _ -> Some e
+    //                    | TField({eexpr = TTypeExpr _},FEnum _) -> Some e
+    //                    | TCast(e,None) -> loop e
+    //                    | _ ->
+    //                         if ctx.com.display.dms_kind = DMNone || ctx.com.display.dms_inline && ctx.com.display.dms_error_policy = EPCollect then
+    //                             display_error ctx "Parameter default value should be constant" p;
+    //                         None
+    //               in
+    //               loop e
+
+    // Maps, Arrays, Function Literals, Object, Anonymous, Block Statements are
+    // Haxe LiteralExpressions.  Booleans, LITINT, LITHEX, LITOCT, LITFLOAT, true, false, null
+    // are either HaxeLiteralExpressions OR HaxeConstantExpressions, depending upon
+    // the rule that they matched in the BNF.
+    if (is_type(expr, HaxeLiteralExpression.class)
+    ||  is_type(expr, HaxeConstantExpression.class)) {
+      PsiElement childElement = expr.getFirstChild();
+      IElementType type = null != childElement ? childElement.getNode().getElementType() : null;
```

Comment:
It would be better to check if type is null after this line. It will reduce amount of unnecessary checks.
`if (type == null) return ConstantClass.NOT_CONSTANT`